### PR TITLE
Add button to allow players to change teams

### PR DESCRIPTION
**Pull request by Zanieon**
This a feature not related to vanilla behavior but rather QoL in regards to the very nature of the Northstar servers, as the only way of changing teams is using mods that gives these functionalities to the server, when it actually should be bundled in Northstar itself already, imo.

image

### CODE REVIEW

**Comment by Zanieon** [comment]
@F1F7Y This one is specially for you due to your Server Utilities :D

**Review by ASpoonPlaysGames — CHANGES_REQUESTED** [review]
If you made the convar replicated, you could enable and disable the UI button based on if the server has the convar enabled. I think we should try to avoid using SendHudMessage for most things if possible

**Comment by Zanieon** [comment]
> If you made the convar replicated, you could enable and disable the UI button based on if the server has the convar enabled. I think we should try to avoid using SendHudMessage for most things if possible

Alright, convar replicated and button visibility bound to it, now in regards to SendHudMessage, those i'll have to probably localize them eventually because would be quite frustrating to players trying to click the button and no info will ever pop in for those two specific cases they could appear.

File: Northstar.Client/mod/scripts/vscripts/ui/menu_ingame.nut
Comment (F1F7Y) [inline Northstar.Client/mod/scripts/vscripts/ui/menu_ingame.nut:91]: Why did you remove this?

File: Northstar.Client/mod/scripts/vscripts/ui/menu_ingame.nut
Comment (F1F7Y) [inline Northstar.Client/mod/scripts/vscripts/ui/menu_ingame.nut:91]: Variable name

File: Northstar.Client/mod/scripts/vscripts/ui/menu_ingame.nut
Comment (ASpoonPlaysGames) [inline Northstar.Client/mod/scripts/vscripts/ui/menu_ingame.nut:89]: When does this code run? It's on the UI VM so I would assume once, in which case this code won't work as you intend.

File: Northstar.Client/mod/scripts/vscripts/ui/menu_ingame.nut
Comment (ASpoonPlaysGames) [inline Northstar.Client/mod/scripts/vscripts/ui/menu_ingame.nut:91]: Why remove this?

File: Northstar.CustomServers/mod/scripts/vscripts/mp/_base_gametype_mp.gnut
Comment (F1F7Y) [inline Northstar.CustomServers/mod/scripts/vscripts/mp/_base_gametype_mp.gnut:646]: Would be great if we didn't mix indent and normal scopes

File: Northstar.CustomServers/mod.json
Comment (F1F7Y) [inline Northstar.CustomServers/mod.json:56]: I know spoon wanted this ( and I agree with it ) but would be great to know version compat of this change before release

**Review by F1F7Y — CHANGES_REQUESTED** [review]
Train WiFi review

File: Northstar.Client/mod/scripts/vscripts/ui/menu_ingame.nut
Comment (ASpoonPlaysGames) [inline Northstar.Client/mod/scripts/vscripts/ui/menu_ingame.nut:91]: I *believe* that there is an overload of this function that can take a callback for if the button is locked or not, might be better to use that and simple disable the button instead of fully hiding it?

File: Northstar.Client/mod/scripts/vscripts/ui/menu_ingame.nut
Comment (Zanieon) [reply Northstar.Client/mod/scripts/vscripts/ui/menu_ingame.nut:91]: Forgot to restore after i started experimenting with the button, in the early attempts.

Comment (Zanieon) [reply Northstar.Client/mod/scripts/vscripts/ui/menu_ingame.nut:91]: Forgot to restore after i started experimenting with the button, in the early attempts.

File: Northstar.Client/mod/scripts/vscripts/ui/menu_ingame.nut
Comment (Zanieon) [reply Northstar.Client/mod/scripts/vscripts/ui/menu_ingame.nut:91]: What about it?

File: Northstar.CustomServers/mod/scripts/vscripts/mp/_base_gametype_mp.gnut
Comment (Zanieon) [reply Northstar.CustomServers/mod/scripts/vscripts/mp/_base_gametype_mp.gnut:646]: Will change then

File: Northstar.CustomServers/mod.json
Comment (Zanieon) [reply Northstar.CustomServers/mod.json:56]: That would be up to @GeckoEidechse when he does a release with this change right?

File: Northstar.Client/mod/scripts/vscripts/ui/menu_ingame.nut
Comment (Zanieon) [reply Northstar.Client/mod/scripts/vscripts/ui/menu_ingame.nut:89]: IIRC Convars replicates from server before map load, though i just found a callback which i could set this up probably.

File: Northstar.Client/mod/scripts/vscripts/ui/menu_ingame.nut
Comment (Zanieon) [reply Northstar.Client/mod/scripts/vscripts/ui/menu_ingame.nut:91]: Yeah gonna use `Hud_SetLocked( devButton, true )` instead of hiding, i actually forgot that's an option for buttons.

Comment (ASpoonPlaysGames) [reply Northstar.Client/mod/scripts/vscripts/ui/menu_ingame.nut:91]: It doesn't make sense, its the "switch teams button" but the variable is called `devButton`

Comment (Zanieon) [reply Northstar.Client/mod/scripts/vscripts/ui/menu_ingame.nut]: Oh yeah stupid of my part, i copy pasted from the DEV block but didn't even notice i could change the var name.

**Review by ASpoonPlaysGames — APPROVED** [review]
Code looks good as far as I'm concerned

**Comment by Alystrasz** [comment]
> In regards to SendHudMessage, those i'll have to probably localize them eventually because would be quite frustrating to players trying to click the button and no info will ever pop in for those two specific cases they could appear.

I agree information should be transmitted to player somehow.
A less invasive way could be to print warning message in chat (using another colour than white).

File: Northstar.Client/mod/scripts/vscripts/ui/menu_ingame.nut
Comment (F1F7Y) [inline Northstar.Client/mod/scripts/vscripts/ui/menu_ingame.nut:707]: Should be suffixed with `_Threaded`

**Review by F1F7Y — CHANGES_REQUESTED** [review]

File: Northstar.Client/mod/scripts/vscripts/ui/menu_ingame.nut
Comment (Zanieon) [reply Northstar.Client/mod/scripts/vscripts/ui/menu_ingame.nut]: Done

**Review by F1F7Y — APPROVED** [review]
I hate the thread thing but if it works it works

**Comment by Zanieon** [comment]
> I hate the thread thing but if it works it works

It's actually not really necessary, but talking with @ASpoonPlaysGames on Discord VC, we agreed that having that threading is good because then Clients updates realtime in case server enables/disables the ability to change teams.

**Comment by F1F7Y** [comment]
I fully agree the button should be up to date, I'm just questioning whether a while true loop is the best way to do it.

**Comment by Zanieon** [comment]
> I'd say that if we can get a convar change callback to handle it then that would be ideal but I dont think we expose that sort of thing to squirrel?

There's callbacks for when networked variables changes in Squirrel, but there's none for convars afaik.

**Comment by Gazyi** [comment]
Changing team should kill player and Auto-Titan in playing state - player can still be near old team.

**Comment by Zanieon** [comment]
> Changing team should kill player and Auto-Titan in playing state.

Why?

**Comment by Gazyi** [comment]
> Why?
For example, you can just switch team and kill old teammates nearby. 
Also, that's how it works in other Source games.

**Comment by Zanieon** [comment]
Valve games don't have Squirrel callbacks for changing teams, so it's easier to just kill the player to fix those potential issues.

FSU existed for a long time now and nobody ever complained about its method of switching teams without killing players, for CTF i can just add a team change callback to force flag drop in case player is carrying it.

**Comment by ASpoonPlaysGames** [comment]
Fairly certain there is already a callback for changing teams it just never gets called? I remember seeing the killfeed code has stuff for team change notifications and somewhere up the callstack for that it comes from a callback?

**Comment by Zanieon** [comment]
> Fairly certain there is already a callback for changing teams it just never gets called? I remember seeing the killfeed code has stuff for team change notifications and somewhere up the callstack for that it comes from a callback?

That's what i mean, adding that callback for special logic in such case since it is already used to notify players when someone changes teams which you are probably remembering from #789

**Review by Alystrasz — APPROVED** [review]
Code looks good, team change is now possible in-game; it's impossible to spam team change more than once every 5 seconds.

File: Northstar.CustomServers/mod.json
Comment (GeckoEidechse) [reply Northstar.CustomServers/mod.json:56]: I don't have the capacity to test that atm. If someone could take care of that, that would great.
It can be tested without making a release.

Basically testing should cover
- New server <-> Old Client
- Old server <-> New Client

And on that note
- New server with old client not working is a non-issue as we can just version gate
- Old server with new client would suck as servers don't update that fast. We could still do it though, just need to ping server hosts.

Comment (Zanieon) [reply Northstar.CustomServers/mod.json:56]: New Clients to Old Servers won't change anything, the server will recieve the command from the button but ignore it.

New Servers to Old Clients will also not change anything substantial besides old clients having to type "changeteam" in console in order to swap.

**Comment by Bobbyperson** [comment]
Might I suggest that this convar be set to zero automatically in some cases? This would mess with infection and the hidden pretty badly.

**Comment by Zanieon** [comment]
> Might I suggest that this convar be set to zero automatically in some cases? This would mess with infection and the hidden pretty badly.

If the server is fixed at those gamemodes then all you need to do is just add the convar to the launch arguments, disabling it, the other method is make the gamemode initialization function disable it as well.

**Comment by Bobbyperson** [comment]
True, some people might not be running dedicated servers though so it might be worthwhile adding it to those functions like you said.

**Comment by Zanieon** [comment]
> True, some people might not be running dedicated servers though so it might be worthwhile adding it to those functions like you said.

Alright i'll probably make a proper toggle function instead of fiddling with the convar for gamemodes that does custom team switching.

**Comment by Zanieon** [comment]
@Bobbyperson See if this is okay for you now.

**Comment by Bobbyperson** [comment]
Looks great, only other changes I might suggest is to also disable it on competitive gamemodes like Fastball and Live fire, since changing teams in those gamemodes could cause big problems.

It might also be worth adding a convar like `ns_max_team_imbalance` (default 1) that would stop specific players (trolls) from switching to prevent teams from having too many/too few players. For example, setting it to 2 would allow a 2 player difference between the teams.

**Comment by Zanieon** [comment]
> @Zanieon what's blocking this PR atm? Just the changes @Bobbyperson suggested?

Nothing really, just pondered a bit if i should implement Bobbyperson suggestion in regards to balance of teams or not, but at that point, trolls should be punishable with either a kick or warn for doing such thing, the intention of the button is not to handle all cases of all types of players.

Fine for me to merge as is, if these cases of people joining one team way too much then it might be worth doing another PR specifically for said case.

**Comment by Bobbyperson** [comment]
AFAIK, most Northstar servers do not have active moderators. In my opinion, if we add a feature like this we should make it as troll proof as possible as to not add an extra burden for server owners/moderators to manage. 

If you don't wanna do it I can make a PR for this later after merge.

**Comment by Zanieon** [comment]
> AFAIK, most Northstar servers do not have active moderators. In my opinion, if we add a feature like this we should make it as troll proof as possible as to not add an extra burden for server owners/moderators to manage.
> 
> If you don't wanna do it I can make a PR for this later after merge.

I'd prefer to see how this will test itself out first without it, so far all servers that allows team switching with FSU never had much issues with that, this button will also probably not bypass the decision made by the Better Team Balance mod (which is another thing that should eventually be integrated into northstar as an option).

**Pull request merged**
